server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -664,30 +677,50 @@ void beginPollingClusterFormationInfo(
      * completed, adding the resulting Cancellable to cancellableConsumer.
      * @param masterEligibleNode The node being polled
      * @param responseConsumer The response consumer to be wrapped
-     * @param cancellableConsumer The list of Cancellables
+     * @param cancellables The Map of Cancellables, one for each node being polled
      * @return
      */
     private Consumer<CoordinationDiagnosticsService.ClusterFormationStateOrException> rescheduleFetchConsumer(
         DiscoveryNode masterEligibleNode,
         Consumer<CoordinationDiagnosticsService.ClusterFormationStateOrException> responseConsumer,
-        Consumer<Scheduler.Cancellable> cancellableConsumer
+        Map<DiscoveryNode, Scheduler.Cancellable> cancellables
     ) {
         return response -> {
-            cancellableConsumer.accept(
-                fetchClusterFormationInfo(
-                    masterEligibleNode,
-                    responseConsumer.andThen(rescheduleFetchConsumer(masterEligibleNode, responseConsumer, cancellableConsumer))
-                )
-            );
+            if (clusterFormationInfoTasks != null) {
+                if (cancellables.equals(clusterFormationInfoTasks)) {
```

Comment:
Shall we document the why behind this if statement? Something along the lines of:

```
// it could happen that the polling for cluster formation info was started (1), cancelled and a new one was started (2) 
// before we got a response from the firstly run (1) polling track
// this checks we are still the running polling track
```